### ISSUE TITLE
TTX extracts axis values with low precision from fvar

### ISSUE THREAD

**Issue by Lorp** [fonttools/fonttools#737]
While experimenting with using TTX and OT variations fonts, I was surprised to see that the axis values in fvar (min, max and default, and also settings for fvar's Named Instances) are extracted with low precision. It turns out it's because of the use of fixedToFloat(value, 16). This converts Fixed (16.16) values into decimal floats, using as few digits as possible, so as to still guarantee that converting back to Fixed will be round-trip safe.

I think this choice is the wrong decision for a tool such as TTX, and for some other uses of fontTools — see the use case below.

First, an axample.

Skia contains the following data (hex, then dex) for min, default, max:

wght: 00007AE1, 00010000, 00033333 = 31457, 65536, 209715
wdth: 00009EB7, 00010000, 00014CCC = 40631, 65536, 85196

These values, if extracted with "double" precision, would allow us to set up two extreme corners of Skia using the following CSS:

font-variation-settings: 'wght' 0.47999572753906, 'wdth' 0.61997985839844;
font-variation-settings: 'wght' 3.1999969482422, 'wdth' 1.2999877929688;

However, TTX extracts the following:

   
   <Axis>
     <AxisTag>wght</AxisTag>
     <MinValue>0.48</MinValue>
     <DefaultValue>1.0</DefaultValue>
     <MaxValue>3.2</MaxValue>
     <NameID>257</NameID>
   </Axis>

   
   <Axis>
     <AxisTag>wdth</AxisTag>
     <MinValue>0.61998</MinValue>
     <DefaultValue>1.0</DefaultValue>
     <MaxValue>1.29999</MaxValue>
     <NameID>258</NameID>
   </Axis>

So, we have used a couple fewer bytes, but in the process have introduced some uncertainties.

Imagine the use case of a UI, or an animation, which called for extremely fine control over the variation axis, more precise than the 1/65536 quantization allows. If the min, max and default values are extracted using TTX, the controls would not work correctly around those values. It seems analogous to control points being on a coarse grid, while the curves that they define follow much more precise positions. Thus – for fvar at least – I would advocate for "double"-style precision in the decimal representation of these values in TTX. I would also suggest checking where else fixedToFloat() is used, and verifying if there is any benefit. Note that fixedToFloat() is a slow function compared to the simple floating point division that it replaces.

One solution for TTX could be to output these values in the form "<n>/65536", where n is the value treated as a LONG, with an XML comment to note the (accurate) floating point value. When compiling, the slash notation would obviously be optional.

Is there a downside to this suggestion?

I believe the code in question is all in:

https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/ttLib/tables/_f_v_a_r.py

**Comment by anthrotype** [comment]
The [`fvar`](https://www.microsoft.com/typography/otspec/fvar.htm) spec uses `Fixed` data type for min, default and max values. In OpenType format, `Fixed` means `32-bit signed fixed-point number (16.16)` (see https://www.microsoft.com/typography/otspec/otff.htm).
The use `fixedToFloat` with 16 decimal places derives from that. You can't get any greater precision without changing the specs.
Using integers in the XML instead of floats won't change anything, as it's still the same number.

```python
>>> floatToFixed(1000.0000499999999, 16)
65536003

>>> floatToFixed(1000.00005, 16)
65536003
```

**Comment by Lorp** [comment]
I understand about the use of Fixed in the spec, which is fine. Yet TTX and fontTools being safe in the round-trip is not proof this is the right approach.

Imagine a user agent that uses TTX to extract the min and max values of Skia’s wght axis, 0x00009EB7 and 0x00014CCC respectively. It obtains 0.61998 and 1.29999. Now imagine another user agent that reads the fvar table directly: it obtains 0.61997985839844 and 1.2999877929688. These are not the same and I think that's a (small) problem for extremes, defaults and named instances. Those two implementations should result in identical behaviour, as far as possible.

In other words, there seems to be no good reason that axis positions in a font system (not a font file) should be quantized at the resolution of 16.16, something that could prevent smooth animations in some circumstances.

**Comment by justvanrossum** [comment]
@Lorp is correct, and we should consider the consequences. It's indeed not so much about `1000.0000499999999` vs. `1000.00005` but rather `0.61998` vs. `0.61997985839844`.

**Comment by justvanrossum** [comment]
Btw. this is not only about axes data, but rather about all Fixed handling in FontTools, anything that goes through `fixedToFloat()`.

**Comment by justvanrossum** [comment]
But not in float!

**Comment by anthrotype** [comment]
if you want "double" precision, ask to change the spec to use 64 bit :)

**Comment by justvanrossum** [comment]
It's not about the spec, it's about how Fixed values are represented as floats in Python.

**Comment by anthrotype** [comment]
A fixed-point number is different from a floating-point number.
We represent it as float in fonttools simply to make it more "user-friendly". But in reality it is a fractional binary number that uses a fixed number of bits.
They are named "fixed" because the so-called "binary point" (somewhat similar to the "decimal point" for base 10 numbers) resides at a fixed location between the bits in the number. 
For example, the `Fixed` data type in OpenType is a signed 32-bit number where the bits are split in half: 16 bits for the whole part, and 16 bits for the fractional part (actually, the most significant bit in the whole number is the "sign" bit).

**Comment by Lorp** [comment]
Just and I both understand Fixed values :) Yet this low-precision approach for *reading* them makes it difficult to write a fontTools-based rasterizer that behaves like one that reads tables directly. To emulate, say, Apple or Microsoft, I would have to convert that 0.61998 back into a Fixed, then use that accurate value (0x00009EB7 / 0x00010000) in my code.

The first version of my Axis-Praxis used TTX to extract values from fvar, then set axis positions uings CSS font-variation-settings. It bugged me that – depending on how variations are implemented in WebKit and macOS – I was not hitting exactly the same axis values as Apple did. I rewrote Axis-Praxis to read fvar directly, and now I am happy.

As @justvanrossum says, this is about more than fvar.

**Comment by Lorp** [comment]
This also affects glyf. Transformations of components defined in the glyf table use F2DOT14. ```decompile()``` is similarly compromised due to its reliance on ```fixedToFloat```. A fontTools-based renderer will not give the same results as other renderers that read the table directly.

```
from fontTools.misc.fixedTools import fixedToFloat as fi2fl, floatToFixed as fl2fi
[...]
		if self.flags & WE_HAVE_A_SCALE:
			scale, = struct.unpack(">h", data[:2])
			self.transform = [[fi2fl(scale,14), 0], [0, fi2fl(scale,14)]]  # fixed 2.14
			data = data[2:]
		elif self.flags & WE_HAVE_AN_X_AND_Y_SCALE:
			xscale, yscale = struct.unpack(">hh", data[:4])
			self.transform = [[fi2fl(xscale,14), 0], [0, fi2fl(yscale,14)]]  # fixed 2.14
			data = data[4:]
		elif self.flags & WE_HAVE_A_TWO_BY_TWO:
			(xscale, scale01,
					scale10, yscale) = struct.unpack(">hhhh", data[:8])
			self.transform = [[fi2fl(xscale,14), fi2fl(scale01,14)],
							[fi2fl(scale10,14), fi2fl(yscale,14)]] # fixed 2.14
			data = data[8:]
```

**Comment by Lorp** [comment]
Cosimo, your comment about 64 bits seems to suggest that we are asking for an arbitrary level of precision. It's not arbitrary though! It's simply what you get from Python (or JavaScript or PHP) if you divide an integer by 65536.0 and print the result. That's the right approach for fontTools decompilation.

**Comment by justvanrossum** [comment]
@anthrotype, please consider the actual problem that Laurence ran into with Axis-Praxis.

**Comment by Lorp** [comment]
If we reduce it to a simple example, let’s say there was a Fixed 6.2 data type, 6 bits of integer and 2 of fraction. Starting from 0, and incrementing by the smallest amount, fixedToFloat(n, 2) gives:

0x00: 0
0x01: 0.2
0x02: 0.5
0x03: 0.8
0x04: 1
…

I suggest that almost all developers would prefer to see 0.25 and 0.75 as the decimal representation of 0x01 and 0x03, instead of 0.2 and 0.8. Imagine going on to use these values in further calculations, and it’s easy to see how errors would rapidly accumulate.

Are there compelling advantages to fixedToFloat that I’m not seeing?

**Comment by justvanrossum** [comment]
@behdad, are you following this? I think https://github.com/fonttools/fonttools/commit/2d67a188b1bef1b70b50b0862a19210d0d4c02f2 was a mistake. It's nice for humans, but it causes actual other problems.

**Comment by behdad** [comment]
Yes I'm following closely.  I sense there's a deep difference of point of view here that cannot easily be reconciled :).

In particular, ' 1.2999877929688 is NOT more precise than ' 1.29999.  The former has 13 digits after decimal point.  To represent 13 decimal digits you need at least 44 bits.  Whereas we know the number is encoded using 16 bits only, so 5 digits is enough.

I understand that Laurence has hit a real problem.  It is my understanding that the issue is a bug in CoreText, which seems to be rejecting settings if they are outside the valid range, instead of clamping to the range.  I understand that "not our bug" is not an acceptable resolution, at least to some of you.  So I've been thinking what we can do to resolve this.  I particular, I don't want to see us reverting fixedToFloat.

Laurence: what you can do to work around this issue, is when you get the number from fonttools, convert it to fix and back to float yourself, ie: `round(f*65536)/65536.`. That gives you the exact number.

Let's think a bit more about this.

**Comment by justvanrossum** [comment]
But 1.2999877929688 *represents* The 16.16 value of 0x00014CCC more precisely than 1.29999. The difference may be tiny in the 16.16 case, but it's already larger in the 2.14 transform values.

**Comment by Lorp** [comment]
@behdad good to see you :)

I did hit a problem in WebKit, but that’s something unrelated: a weird issue such that values near (but within) the extremes are sometimes interpreted as beyond the range, and hence (as is intended behaviour for values outside the range) revert to default for that axis. This was all using CSS ```font-variation-settings```, so it’s not necessarily deep in CoreText.

Back to our issue.

You say "To represent 13 decimal digits you need at least 44 bits". This is true in the general case, but I could equally say "To represent 2 binary fractional digits we need 2 decimal fractional digits" (as in my hypothetical Fixed 6.2 example above) — you’ve got this "need" the wrong way round IMO. We are reading data in decompile, and (again from my simplified example) "0.75" is a much better decimal reading of 0x03 than "0.8".

You also say that "we know the number is encoded using 16 bits only". That is true, spec geeks know this, but TTX users should not have to!

**Comment by behdad** [comment]
I think I agree that we shouldn't round when loading, but when printing to XML.

**Comment by justvanrossum** [comment]
I would have far less problems with that option! (Rounding when writing to XML.)

**Comment by behdad** [comment]
> and hence (as is intended behaviour for values outside the range) revert to default for that axis.

That's what CoreText does now.  But OT 1.8 requires clamping, not ignoring.

**Comment by justvanrossum** [comment]
A subclass of float could carry info about the original precision, so that could be a solution.

**Comment by behdad** [comment]
> and (again from my simplified example) "0.75" is a much better decimal reading of 0x03 than "0.8".

The 2bit case is an exaggeration.  Let's stick to the real cases :).

**Comment by justvanrossum** [comment]
The 2bit case is an excellent explanation of why the current approach is flawed.

**Comment by justvanrossum** [comment]
Something like this could be useful if we only want to round when writing to XML:
```
class precision_float(float):
    def __new__(cls, value, precision=None):
        return super(precision_float, cls).__new__(cls, value)
    def __init__(self, value, precision=None):
        super(precision_float, self).__init__(value)
        self.precision = precision
    def __str__(self):
        if self.precision is None:
            return super(precision_float, self).__str__()
        return str(round(self, self.precision))

f1 = 3.45634287984792348

f2 = precision_float(f1, 5)

print(f1)
print(repr(f1))
print(f2)
print(repr(f2))
```

**Comment by Lorp** [comment]
(About CoreText clamping: Apple knows about this and, I believe, is updating old code that was intended not to clamp, to conform to the new spec.)

**Comment by Lorp** [comment]
There’s a comparable case in digital mapping. Many coordinates were recorded originally in degrees, minutes and seconds. Thus one might find in a text or a database that KABK is located at 52°4′53″N, 4°19′5″E. Of course these days it’s more common to express things in decimal degrees: 52.081389°, 4.318056°. On Google Maps these point to exactly the same place.

It’s interesting to know what the resolutions of these two systems are, so let’s calculate them (in the north-south direction).

In the degrees-arcminutes-arcseconds system, the smallest unit, one arcsecond, is 1/60 of a nautical mile, or 30.866666666666667 m.

In the decimal system, the smallest unit is 1×10⁻⁶ of a degree, or 6×10⁻⁵ nautical miles, or 0.11112 m.

So we have a situation where the sources of many coordinates are in a much coarser resolution, 277.777777778 times coarser to be precise, than the resolution in general use. I think you can see where I am going… Except on a display with limited space, or to an audience who might find the long numbers overwhelming, no modern digital mapper would ever truncate decimal values after conversion!

**Comment by tshinnic** [comment]
Not meaning to "pile on" and this might just confuse things more, but...

TL;DR - F2DOT14 results do not 'match' three OTF spec'd example values.

I was reviewing the specs and getting confused by this code and the various issues (past and present) here, and then realized that `Tests/misc/fixedTools_test.py` does not include the [examples from the spec](https://www.microsoft.com/typography/otspec/otff.htm) (find-in-page for "9999").  Those examples are:
```
  Decimal Value   Hex Value   Integer    Fraction
     1.999939       0x7fff       1      16383/16384
     1.75           0x7000       1      12288/16384
     0.000061       0x0001       0          1/16384
     0.0            0x0000       0          0/16384
    -0.000061       0xffff      -1      16383/16384
    -2.0            0x8000      -2          0/16384
```
Those would translate into

        self.assertEqual(1.999939,  fixedToFloat(32767, 14))    # 0x7FFF
        self.assertEqual(1.75,      fixedToFloat(28672, 14))    # 0x7000
        self.assertEqual(0.000061,  fixedToFloat(0x0001, 14))
        self.assertEqual(0.0,       fixedToFloat(0x0000, 14))
        self.assertEqual(-0.000061, fixedToFloat(-1, 14))       # 0xFFFF
        self.assertEqual(-2.0,      fixedToFloat(-32768, 14))   # 0x8000

and I added for completeness:

        self.assertEqual(-1.999939, fixedToFloat(-32767, 14))   # 0x8001
        self.assertEqual(1.000061,  fixedToFloat(-16385, 14))   # 0xBFFF

I bet you can guess which ones 'fail' ...

       E       AssertionError: 1.999939 != 1.99994
       E       AssertionError: 6.1e-05 != 6e-05
       E       AssertionError: -6.1e-05 != -6e-05
       E       AssertionError: -1.999939 != -1.99994
       E       AssertionError: 1.000061 != -1.00006

The actual values I get using plain math for these are:

    v: 0x7fff  result:  1.99993896484375
    v: 0x0001  result:  0.00006103515625
    v: 0xffff  result: -0.00006103515625

    v: 0x8001  result: -1.99993896484375
    v: 0xbfff  result: -1.00006103515625

**Comment by anthrotype** [comment]
I think the consensus here was we shouldn't round when loading, only when printing to XML.
Any volunteers?

**Comment by justvanrossum** [comment]
@Lorp's point bears repeating (about F2DOT14 transformation vaues):

>  A fontTools-based renderer will not give the same results as other renderers that read the table directly.

I think that sucks big time. Brings it back to #1204 :)

**Comment by anthrotype** [comment]
ok, so.
We should add an option like `shortest=False` to `fixedToFloat` function, so that whenever it's used inside `decompile` method in ttLib tables, it will just `return value / (1 << precisionBits)`.

Then, from within the `toXML` methods we need to convert the float (it's no longer a fixed by then) to it's shortest representation, so we need to do `fixedToFloat(floatToFixed(key, 14), 14, shortest=True)`.

Then, from within `fromXML` methods, we need to quantize the float to F2Dot14 as in `floatToFixedToFloat` (which is the same as  `fixedToFloat(floatToFixed(key, 14), 14, shortest=False)`).

Correct?

**Comment by justvanrossum** [comment]
I'd say that `fixedToFloat()` should _never ever_ perform decimal rounding, and that there should be a decimally optimized fixed repr func of some sort.

Probably the most practical problem is to extend the fixed precision knowledge to the text read/write logic. #492 / #494 are also affected.

**Comment by anthrotype** [comment]
> fixedToFloat() should never ever perform decimal rounding

that's why I proposed a `shortest=False` keyword argument, so by default it would no longer round to shortest. 

> extend the fixed precision knowledge to the text read/write logic. 

what do you mean exactly?

**Comment by justvanrossum** [comment]
> >    fixedToFloat() should never ever perform decimal rounding

> that's why I proposed a shortest=False keyword argument, so by default it would no longer round to shortest.

And I meant what I said, that `fixedToFloat()` should just not round _ever_, not even optionally. Making a (shortest) decimal representation should result in a string, not a rounded-by-decimals-but-represented-as-a-binary-float-which-maybe-later-will-be-converted-to-a-string. So I suggest to have a completely _separate_ function that returns a decimally-rounded string for dumping purposes. The opposite would then also be handy.

> >  extend the fixed precision knowledge to the text read/write logic.

> what do you mean exactly?

I'm just speculating, but there may be places where read/write code is generic, and there's no direct access to the original fixed precision anymore. Either way, one needs to specify the precision in _two_ places instead of one. (Hence my earlier suggestion to use a float subclass, but that doesn't solve the reading problem.)

**Comment by anthrotype** [comment]
ok, got it. Though I'd like to think this through a little bit. Maybe we can move this to the release following the one I'm preparing today.

> a float subclass

i believe that would slow down math operations because instead of an optimized built-in python would have to treat it like a generic object.

**Comment by justvanrossum** [comment]
> Maybe we can move this to the release following the one I'm preparing today.

Certainly. Better do this right than in a hurry!

**Comment by behdad** [comment]
So, the actual fix is not too bad if someone wants to give it a try.  `fixedToFloat()` should be changed to NOT do any magic. `fixedToStr()` and `strToFixed()` and `strToFixedToFloat()` should be added that do magic. `fixedToStr()` will do what fixedToFloat() currently does, except for returning a string. `strToFixed()` then just does `floatToFixed(float(s))` and  indeed `strToFixedToFloat()` does `fixedToFloat(floatToFixed(float(s)))`.

The harder part is fixing all the call sites. For the otData-tables, it's easy: in otConverters.py, the types `Fixed` and `F2Dot14` need to implement `toXML` / `fromXML` that calls the new functions.

The other tables (`avar`, `glyf`, `kern`, `trak`) as well as `TupleVariations` and `psCharStrings` need to be reviewed and fixed.

**Comment by behdad** [comment]
cc @robmck-ms
